### ISSUE TITLE
[HOLD for payment 2023-08-21] [$1000] Multiline codeblocks disappear after editing twice

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#18928]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open the app
2. Open any report
3. Send 3 backtick mutline code block
4. Edit the code block to remove space before last and first characters 
5. Repeat step 4 again

## Expected Result: 
Codeblock should not disappear on edit
## Actual Result: 
Codeblock disappears on edit in mac chrome/mac safari

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.13.3
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/2c81fe08-51d2-492e-b933-db47f1c522ce

https://github.com/Expensify/App/assets/43996225/a5a3981f-0ae0-42f9-9d0d-ebfdc312fd8f

**Expensify/Expensify Issue URL:**
**Issue reported by:** @dhanashree-sawant
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1683890048581339

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01bba6b459f7297429</li>
        <li>Upwork Job ID: 1664260132903624704</li>
        <li>Last Price Increase: 2023-06-08</li>
    </ul>
</details>

**Comment by laurenreidexpensify** [comment]
![2023-05-15_10-36-37 (1)](https://github.com/Expensify/App/assets/62073721/1b64586f-6c29-400e-a549-c37aaa50bdcd)

I can't reproduce this on staging on web

**Comment by laurenreidexpensify** [comment]
Closing as not reproducible

**Comment by dhanashree-sawant** [comment]
Not able to login on web staging but made a video using mobile chrome, was able to reacreate in first edit if kept the first character and last character or word besides the 3 backticks. The video is over 10 mb so will post it on slack in the report [thread](https://expensify.slack.com/archives/C049HHMV9SM/p1683890048581339)

**Comment by laurenreidexpensify** [comment]
Reopening to review and see if still reproducible as per feedback from @dhanashree-sawant

**Comment by eh2077** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Multiline codeblocks disappear after editing twice.

### What is the root cause of that problem?

Adding a comment

````
```
1
2```
````

is translated into html

```html
<pre>1<br />2</pre>
```

The html above is translated into markdown

````
```
1
2
```
````

when click edit the comment.

If edit the comment into 

````
```
1
2```
````

then in method [editReportComment](https://github.com/Expensify/App/blob/98ae1aed1147d6d5875ed9b1bc06871d1a7f4b0f/src/libs/actions/Report.js#L921) 

1. `originalCommentHTML` is equal to

```
<pre>1<br />2</pre>
```

2. `htmlForNewComment.trim()` is 

```html
<pre>1<br />2</pre>
```

3. But `parsedOriginalCommentHTML` is equal to

```html
<pre>1<br />2<br /></pre>
```

after doing `html-to-markdown` and `markdown-to-html` conversion, see

https://github.com/Expensify/App/blob/98ae1aed1147d6d5875ed9b1bc06871d1a7f4b0f/src/libs/actions/Report.js#L935

So, the condition 

https://github.com/Expensify/App/blob/98ae1aed1147d6d5875ed9b1bc06871d1a7f4b0f/src/libs/actions/Report.js#L945

return false and we save the save html to the backend. Note that the backend will return empty message which make the comment disappear

<img width="710" alt="image" src="https://github.com/Expensify/App/assets/117511920/897d48fd-39b6-497b-adf5-f1d9d0cd8e69">

So the root cause is that `parsedOriginalCommentHTML` is not equal to `originalCommentHTML` when editing codeblock by removing the last newline.

### What changes do you think we should make in order to solve the problem?

There'er two potential ways to fix this issue

1. Remove outdated code snippet about the conversions of `parsedOriginalCommentHTML`

https://github.com/Expensify/App/blob/98ae1aed1147d6d5875ed9b1bc06871d1a7f4b0f/src/libs/actions/Report.js#L933-L936

Then we won't save the same html to the backend.

2. Fix it from Expensify-common repo to make sure we get same html of codeblock after doing `html-to-markdown` and `markdown-to-html` conversion. This solution could be tricky compared to the first solution.

### What alternative solutions did you explore? (Optional)

N/A

**Comment by blockchainGuru1018** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Multiline codeblocks disappear after editing twice.

### What is the root cause of that problem?

Regarding the add a comment


When you edit the comment like the bellow

````
```
1
2```
````

then in method [editReportComment](https://github.com/Expensify/App/blob/98ae1aed1147d6d5875ed9b1bc06871d1a7f4b0f/src/libs/actions/Report.js#L921) 

`originalCommentHTML` and `htmlForNewComment.trim()` are equal to

```
<pre>1<br />2</pre>
```

But `parsedOriginalCommentHTML` is

```html
<pre>1<br />2<br /></pre>
```

As a result, this condition returning false value

https://github.com/Expensify/App/blob/98ae1aed1147d6d5875ed9b1bc06871d1a7f4b0f/src/libs/actions/Report.js#L945

That's why the backend will return empty value for the `text` in their response.

![image](https://github.com/Expensify/App/assets/58952144/b8f27972-42a8-4999-96b3-2a2b2db09158)


But when we switch to other tabs and back to this tab, it corrects with `OpenReport` query
![image](https://github.com/Expensify/App/assets/58952144/7ed91f53-e1e1-4810-a575-56fc61aa74c0)



### What changes do you think we should make in order to solve the problem?

Remove outdated code snippet about the conversions of `parsedOriginalCommentHTML`

https://github.com/Expensify/App/blob/98ae1aed1147d6d5875ed9b1bc06871d1a7f4b0f/src/libs/actions/Report.js#L933-L936

Then we won't save the same html to the backend.

### What alternative solutions did you explore? (Optional)

N/A

**Comment by laurenreidexpensify** [comment]
@Santhosh-Sellavel bump for review on these ^^

**Comment by Santhosh-Sellavel** [comment]
Unassigned this due to low bandwidth, can you assign a new C+ here by applying `external` label thanks!

@puneetlath @laurenreidexpensify

**Comment by laurenreidexpensify** [comment]
@abdulrahuman5196 can you pick up review here? thanks

**Comment by laurenreidexpensify** [comment]
@maddylewis I am OOO til 19 June, adding you to keep an eye on til I am back. Thanks

**Comment by abdulrahuman5196** [comment]
@laurenreidexpensify  thanks for the ping. Will close out before EOD.

**Comment by abdulrahuman5196** [comment]
@eh2077 From Reviewing proposal here - https://github.com/Expensify/App/issues/18928#issuecomment-1572262676
I am not exactly sure of the root case. I agree its partially correct but I am not aligned with the issue fix path.

you had mentioned

>we save the same html to the backend. Note that the backend will return empty message which make the comment disappear

Shouldn't the question be why does backend return empty message(Even though the message is similar)? 

@puneetlath  I think backend is returning empty response when the comment is edited the second time in the mentioned way. Could you kindly check on what the issue in backend is or is it like backend doesn't expect similar messages? 

Note: I am not totally aligned to fixing this issue since its super edge case, but curious to check on backend why this issue is happening since it could lead to other issues if not checked

<img width="425" alt="Screenshot 2023-06-07 at 11 31 13 PM" src="https://github.com/Expensify/App/assets/46707890/38f8fc57-2471-47b6-ad6b-55b995752755">

**Comment by eh2077** [comment]
@abdulrahuman5196 Thanks for reviewing my proposal! 

> you had mentioned
> 
> > we save the same html to the backend. Note that the backend will return empty message which make the comment disappear
> 
> Shouldn't the question be why does backend return empty message(Even though the message is similar)?

That's a good question. Yes, I also agree returning empty data from backend looks weird. While what the backend should return if frontend posts the same data without change? I think the backend can consider to return error with proper http code. I found a related [SO discussion](https://stackoverflow.com/questions/3825990/http-response-code-for-post-when-resource-already-exists) FYI. 

I think we'll also need frontend fix to avoid saving same `html` message to the backend, so I focused more on the solution from frontend in my proposal. I have noticed this backend behaviour in another issue which ended up fixing from frontend, see https://github.com/Expensify/App/issues/17785#issuecomment-1536639742

cc @puneetlath

**Comment by abdulrahuman5196** [comment]
@puneetlath Waiting for you on this https://github.com/Expensify/App/issues/18928#issuecomment-1581282296

**Comment by puneetlath** [comment]
Hmm, interesting. To make sure I understand. When a code-block is edited to have the exact same value as before, the back-end returns an empty response? And this is only for code-blocks, not any other type of message?

**Comment by eh2077** [comment]
@puneetlath I can confirm that this happens for other html messages too. I noticed the behavior of backend when I worked on other similar comment disappearing issues previously.

**Comment by abdulrahuman5196** [comment]
Hey folks, @puneetlath is out this week I think. Can we apply labels to assign another internal engineer?

**Comment by abdulrahuman5196** [comment]
@laurenreidexpensify / @maddylewis puneetlath@ is out this week, can we check with internal engineer as per new process for this question?

The issue also changed internal meanwhile

https://github.com/Expensify/App/issues/18928#issuecomment-1581282296

**Comment by laurenreidexpensify** [comment]
Sorry @abdulrahuman5196 picking this up again - @puneetlath is also back from OOO tomorrow so going to leave assigned to him. I'm also overriding that this is an internal issue, I still think it's fine for it to be external. 

Can you confirm what the outstanding questions we have here are? Are we happy with any of the proposals above?

**Comment by abdulrahuman5196** [comment]
@laurenreidexpensify This issue is mostly inclined towards a backend/internal fix as per here https://github.com/Expensify/App/issues/18928#issuecomment-1581282296. We are waiting on @puneetlath to check from backend.

**Comment by puneetlath** [comment]
I will try to take this to the finish line next week after SNH.

**Comment by laurenreidexpensify** [comment]
@puneetlath is gonna take this one after SNH, not overdue

**Comment by puneetlath** [comment]
Started looking into it. Should have an update shortly.

**Comment by laurenreidexpensify** [comment]
@puneetlath is there an update yet? thanks

**Comment by eh2077** [comment]
Friendly bump @puneetlath on https://github.com/Expensify/App/issues/18928#issuecomment-1581282296 and https://github.com/Expensify/App/issues/18928#issuecomment-1581641258. Can you help us move this issue forward?

**Comment by laurenreidexpensify** [comment]
@puneetlath bump ^^

**Comment by Skalakid** [comment]
Hey, I’m Michał from Software Mansion, an expert agency, and I will help close this issue out.

**Comment by Skalakid** [comment]
Is this issue still reproducible on the latest main branch? Because I follow all the issue descriptions and videos and code blocks are not disappearing :( Maybe the bug is already fixed on the backend side. @laurenreidexpensify @eh2077 how about you? Can you still reproduce this bug?

**Comment by eh2077** [comment]
@Skalakid Thanks for checking this. I just tested and it is still reproducible. See below video FYI
https://github.com/Expensify/App/assets/117511920/d13ef634-90ae-4fd1-bc52-424d7af6af07

**Comment by Skalakid** [comment]
Since @puneetlath is not responding, I assume the backend issue is more complicated and I think we should focus now on fixing this bug on the frontend side. 
@eh2077 Thanks for the proposal, I'm not sure about the first solution, how do you know that the snipped you mentioned is outdated and we can remove it? Won't it break other functionalities?

**Comment by abdulrahuman5196** [comment]
>Since @puneetlath is not responding, I assume the backend issue is more complicated and I think we should focus now on fixing this bug on the frontend side.

@Skalakid I think we should first confirm from backend. Maybe with other engineer, if @puneetlath  is unavailable? @laurenreidexpensify

**Comment by Skalakid** [comment]
@abdulrahuman5196 Yes, we can confirm from the backend, however, I think that in the end, the frontend fix is needed. We shouldn't send the server edit request with the same message as it was before editing. So we can check the backend and fix the message checking on the frontend separately at the same time, to move this issue forward. What do you think about it?

**Comment by Skalakid** [comment]
@eh2077 that makes sense. What do you think about it @abdulrahuman5196? 
I was thinking about the solution to this issue, but to be more secure I propose adding `originalCommentHTML === htmlForNewComment.trim()` as an alternative condition:

```
    if (parsedOriginalCommentHTML === htmlForNewComment.trim() || originalCommentHTML === htmlForNewComment.trim()) {
        return;
    }
``` 

However if `parsedOriginalCommentHTML` is outdated I think we can change it to just `originalCommentHTML === htmlForNewComment.trim()`

**Comment by laurenreidexpensify** [comment]
@puneetlath is OOO this week, so will pick up again next week - I will make this a weekly in interim

**Comment by Skalakid** [comment]
@puneetlath any update on this?

**Comment by puneetlath** [comment]
Ok first off, my apologies for my long delay in handling this issue. Secondly, thank you so much to everyone that put in effort looking for a solution. 

I spent a good chunk of time looking into what was going on in the back-end for this today and it turns out the back-end just never really considers that the "edit" might not actually be an edit. 

So I opened a simple back-end PR to handle that scenario: https://github.com/Expensify/Web-Expensify/pull/38307

However, I agree that it is weird that we are sending this to the back-end at all, so I'm down with the front-end solution proposed [here](https://github.com/Expensify/App/issues/18928#issuecomment-1637884028) in addition to the back-end fix if we are confident it will handle our current bug without creating any regressions. 

Thanks again y'all!

**Comment by Skalakid** [comment]
Okay, thank you @puneetlath for investigating this issue on the backend and creating PR with the fix! :D @abdulrahuman5196 what do you think about frontend changes? Should I create PR?

**Comment by abdulrahuman5196** [comment]
@Skalakid  i think there are some frontend proposals left for review. If they doesn't come to the mark. You could start. Will update after review

**Comment by Skalakid** [comment]
@blockchainGuru1018 thanks for the proposal however, @eh2077 was first with such a solution. Your proposals are really similar

**Comment by Skalakid** [comment]
@abdulrahuman5196 I think all proposals are reviewed. What you thoughts on https://github.com/Expensify/App/issues/18928#issuecomment-1637884028?

**Comment by puneetlath** [comment]
@abdulrahuman5196 bump!

**Comment by abdulrahuman5196** [comment]
Sorry this issue had slipped my radar. Mostly likely due to weekly label. Will check and update in couple of hours.

**Comment by abdulrahuman5196** [comment]
Reviewing now

**Comment by abdulrahuman5196** [comment]
Sorry was caught up in deployment blocker PR inbetween. Reviewing again now

**Comment by abdulrahuman5196** [comment]
I think we shouldn't remove the code as suggested here https://github.com/Expensify/App/issues/18928#issuecomment-1572262676. I checked and there are mulitple flows present there and it would be very difficult to say its outdated or not.

I agree on the comment here - https://github.com/Expensify/App/issues/18928#issuecomment-1637884028, we can add a another check of `originalCommentHTML === htmlForNewComment.trim()`. It is harmless and for our case of second edit it solves the purpose.
I am assuming the above proposal from @Skalakid her https://github.com/Expensify/App/issues/18928#issuecomment-1637884028 came from @eh2077 's comment here - https://github.com/Expensify/App/issues/18928#issuecomment-1634404003(Not a formal proposal though)

So I am fine with approving @eh2077 as the contributor. @eh2077  Can you include the same in your original proposal?

Note: @Skalakid Do let me know if my statement about your proposal coming from @eh2077 comment is wrong. And I can revert.

🎀 👀 🎀
C+ Reviewed

cc: @puneetlath

**Comment by eh2077** [comment]
@abdulrahuman5196 thanks for reviewing. I think the selected solution, adding the alternative condition, is from team work with @Skalakid . So I'm open to split the bounty with @Skalakid , like 50/50

**Comment by abdulrahuman5196** [comment]
@eh2077 @Skalakid is from expert contributors so they are paid differently and won't be applicable for splits.

**Comment by puneetlath** [comment]
Cool, sounds good to me. @eh2077 let's make sure we add some automated tests for the various scenarios related to this.

**Comment by Skalakid** [comment]
@abdulrahuman5196 thank you for the review. I created my proposal but before proposing it I wanted to make sure that my solution isn't redundant. Our solutions with @eh2077 are really similar so I decided to propose these changes to @eh2077's idea. So yes, it was cooperation :D

**Comment by eh2077** [comment]
@abdulrahuman5196 I updated the solution to [proposal](https://github.com/Expensify/App/issues/18928#issuecomment-1572262676). 

I also found the bug is a slight different from the bug report now. The edited comment doesn’t disappear but it'll return to unchanged after switching tab.
See below video 

https://github.com/Expensify/App/assets/117511920/2c813d7b-306f-47d8-876c-8ad19569c99a

The difference is that the backend now returns empty `onyxData` data.
<img width="854" alt="image" src="https://github.com/Expensify/App/assets/117511920/d7867f6b-6d9c-4a45-8481-b218a096a21b">

 So the `(edited)` will show briefly as set edited = true here
https://github.com/Expensify/App/blob/5775b4f6ed571496b87a9d85542e6ef716ee5375/src/libs/actions/Report.js#L1074

I think our solution is still applicable to fix the bug.

**Comment by puneetlath** [comment]
Yes, I fixed the back-end response so that it wouldn't disappear. But I agree, we can still fix the front-end so that it isn't considered to be "edited" and the API doesn't even get called in this scenario.

**Comment by eh2077** [comment]
Gentle bump @abdulrahuman5196. The PR https://github.com/Expensify/App/pull/24205 is ready for review, thanks!

**Comment by eh2077** [comment]
@laurenreidexpensify Can you kindly send me the offer in Upwork?

**Comment by dhanashree-sawant** [comment]
Hi @laurenreidexpensify, I think offer sending for this job is left, I too don't have any offer. Can you check once?

**Comment by laurenreidexpensify** [comment]
Folks - massive apologies, I don't know why I thought this had been accepted by everyone last week!! I have sent offers to everyone in Upwork now for job https://www.upwork.com/en-gb/jobs/~0189f147050f87dc01, as the original job here expired. Thanks

**Comment by eh2077** [comment]
@laurenreidexpensify Accepted the offer, thank you!

**Comment by laurenreidexpensify** [comment]
Everyone has been paid in Upwork - @abdulrahuman5196 can you please complete the checklist, thanks
